From 232f30dc50af9a59f78564a19798200bfc7f5919 Mon Sep 17 00:00:00 2001
From: Mohsen Dolaty <mohsen.dolaty@amd.com>
Date: Mon, 6 Mar 2023 15:49:55 -0600
Subject: [PATCH 1/1] amd-apml: select APML over i2c/i3c

Add ability for the user to select running
APML over i2c or i3c bus

Signed-off-by: Mohsen Dolaty <mohsen.dolaty@amd.com>
---
 scripts/set-apml.sh | 81 +++++++++++++++++++++++++++++----------------
 1 file changed, 52 insertions(+), 29 deletions(-)

diff --git a/scripts/set-apml.sh b/scripts/set-apml.sh
index 5c9d7d9..f45c345 100644
--- a/scripts/set-apml.sh
+++ b/scripts/set-apml.sh
@@ -9,12 +9,7 @@ I2C_BUS=2
 I2C_MUX=0x70

 DEV_I2C_SBTSI_PATH="/sys/bus/i2c/drivers/sbtsi"
-DEV_I2C_SBTSI_2="4-22400000001"
-DEV_I2C_SBTSI_2="5-22400000001"
-
 DEV_I2C_SBRMI_PATH="/sys/bus/i2c/drivers/sbrmi"
-DEV_I2C_SBRMI_3="4-22400000002"
-DEV_I2C_SBRMI_3="5-22400000002"

 # i3c bus
 I3C_TOOL="/usr/bin/i3ctransfer"
@@ -37,6 +32,7 @@ DEV_I3C_SBRMI_5="5-22400000002"
 set_i2c_apml()
 {
     echo "Setting I2C-2 Mux for CPU APML "
+    I2C_BUS=2
     i2cset -f -y $I2C_BUS $I2C_MUX 0x46 0x01 || retval=$?
     if [[ "$retval" -ne 0 ]]; then
         echo "Error: Setting I2C Mux Reg 0x46"
@@ -47,7 +43,7 @@ set_i2c_apml()
     fi
     i2cset -f -y $I2C_BUS $I2C_MUX 0x41 0x40 || retval=$?
     if [[ "$retval" -ne 0 ]]; then
-        echo "Error: Setting I2C Mux Reg 0x46"
+        echo "Error: Setting I2C Mux Reg 0x41"
     fi
     if [ -a "$DEV_I2C_3" ]
     then
@@ -63,7 +59,35 @@ set_i2c_apml()
         fi
         i2cset -f -y $I2C_BUS $I2C_MUX 0x41 0x40 || retval=$?
         if [[ "$retval" -ne 0 ]]; then
-            echo "Error: Setting I2C Mux Reg 0x46"
+            echo "Error: Setting I2C Mux Reg 0x41"
+        fi
+    fi
+}
+
+# unbind i2c drivers
+unbind_i2c_drivers()
+{
+    echo "Clearing I2C-2 Mux for CPU APML "
+    I2C_BUS=2
+    i2cset -f -y $I2C_BUS $I2C_MUX 0x40 0x00 || retval=$?
+    if [[ "$retval" -ne 0 ]]; then
+        echo "Error: Setting I2C Mux Reg 0x40"
+    fi
+    i2cset -f -y $I2C_BUS $I2C_MUX 0x41 0x00 || retval=$?
+    if [[ "$retval" -ne 0 ]]; then
+        echo "Error: Setting I2C Mux Reg 0x41"
+    fi
+    if [ -a "$DEV_I2C_3" ]
+    then
+        echo "Clearing I2C-3 Mux for CPU APML "
+        I2C_BUS=3
+        i2cset -f -y $I2C_BUS $I2C_MUX 0x40 0x00 || retval=$?
+        if [[ "$retval" -ne 0 ]]; then
+            echo "Error: Setting I2C Mux Reg 0x40"
+        fi
+        i2cset -f -y $I2C_BUS $I2C_MUX 0x41 0x00 || retval=$?
+        if [[ "$retval" -ne 0 ]]; then
+            echo "Error: Setting I2C Mux Reg 0x41"
         fi
     fi
 }
@@ -160,19 +184,6 @@ set_i3c_apml()
     fi
 }

-# unbind i2c drivers
-unbind_i2c_drivers()
-{
-    # Unbind sbtsi and sbrmi drivers
-    echo  $DEV_I2C_SBTSI_2 > $DEV_I2C_SBTSI_PATH/unbind
-    echo  $DEV_I2C_SBRMI_2 > $DEV_I2C_SBRMI_PATH/unbind
-    if [[ $num_of_cpu == 2 ]]
-    then
-        echo  $DEV_I2C_SBTSI_3 > $DEV_I2C_SBTSI_PATH/unbind
-        echo  $DEV_I2C_SBRMI_3 > $DEV_I2C_SBRMI_PATH/unbind
-    fi
-}
-
 # unbind i3c drivers
 unbind_i3c_drivers()
 {
@@ -192,21 +203,33 @@ unbind_i3c_drivers()
 # check num of cpu
 echo "Num of CPU " $num_of_cpu

-if [ $1 == "bind" ]
-then
-    echo "Bind Drivers"
-    if [ -a "$DEV_I2C_2" ]
-    then
+if [[ $1 == "i3c" ]];then
+    fw_setenv apml_i3c true
+    reboot now
+elif [[ $1 == "i2c" ]];then
+    fw_setenv apml_i3c false
+    reboot now
+elif [[ $1 == "bind" ]];then
+    if [[ -a "$DEV_I2C_2" ]];then
+        echo "Bind I2C Drivers"
         set_i2c_apml
     else
+        echo "Bind I3C Drivers"
         set_i3c_apml
     fi
-else
-    echo "Unbind Drivers"
-    if [ -a "$DEV_I2C_2" ]
-    then
+elif [[ $1 == "unbind" ]];then
+    if [[ -a "$DEV_I2C_2" ]];then
         unbind_i2c_drivers
     else
+        echo "UnBind I3C Drivers"
         unbind_i3c_drivers
     fi
+else
+    echo "set-apml.sh takes a single string as input"
+    echo "    i3c    = set APML to I3C mode, and Reboot"
+    echo "    i2c    = set APML to I2C mode, and Reboot"
+    echo "    bind   = Bind   I2C/I3C drivers for APML"
+    echo "    unbind = Unbind I2C/I3C drivers for APML"
+    echo "APML mode is"
+    fw_printenv |grep apml_i3c
 fi
--
2.25.1
